core/drivers/crypto/hisilicon/hpre_main.c

```diff
@@ -0,0 +1,339 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2023 HiSilicon Limited.
+ * Kunpeng hardware accelerator hpre module init.
+ */
+#include "hpre_main.h"
+
+/* base config */
+#define HPRE_COMMON_CNT_CLR_CE 0x301000
+#define HPRE_CFG_AXCACHE 0x301010
+#define HPRE_RDCHN_INI_CFG 0x301014
+#define HPRE_BD_ENDIAN 0x301020
+#define HPRE_ECC_BYPASS 0x301024
+#define HPRE_POISON_BYPASS 0x30102c
+#define HPRE_BD_ARUSR_CFG 0x301030
+#define HPRE_BD_AWUSR_CFG 0x301034
+#define HPRE_TYPES_ENB 0x301038
+#define HPRE_DATA_RUSER_CFG 0x30103c
+#define HPRE_DATA_WUSER_CFG 0x301040
+#define HPRE_HAC_INT_MASK 0x301400
+#define HPRE_RAS_ECC_1BIT_TH 0x30140c
+#define HPRE_RAS_CE_ENB	0x301410
+#define HPRE_RAS_NFE_ENB 0x301414
+#define HPRE_RAS_FE_ENB	0x301418
+#define HPRE_HAC_INT_SRC 0x301600
+#define HPRE_RDCHN_INI_ST 0x301a00
+#define HPRE_OOO_SHUTDOWN_SEL 0x301a3c
+#define HPRE_CORE_ENB 0x302004
+#define HPRE_CORE_INI_CFG 0x302020
+#define HPRE_CORE_INI_STATUS 0x302080
+/* clock gate */
+#define HPRE_CLKGATE_CTL 0x301a10
+#define HPRE_PEH_CFG_AUTO_GATE 0x301a2c
+#define HPRE_CLUSTER_DYN_CTL 0x302010
+#define HPRE_CORE_SHB_CFG 0x302088
+#define HPRE_CORE_GATE_ENABLE GENMASK_32(31, 30)
+
+#define HPRE_AXCACHE_MASK 0xff
+#define HPRE_HAC_INT_DISABLE 0x1ffffff
+#define HPRE_RAS_CE_MASK 0x1
+#define HPRE_RAS_NFE_MASK 0x1fffffe
+#define HPRE_RAS_FE_MASK 0
+#define HPRE_BD_LITTLE_ENDIAN 0
+#define HPRE_RSA_ENB BIT(0)
+#define HPRE_ECC_ENB BIT(1)
+#define HPRE_BD_ARUSR_MASK 0x2
+#define HPRE_BD_AWUSR_MASK 0x102
+#define HPRE_DATA_USR_MASK 0x32
+#define HPRE_CLUSTER_CORE_MASK GENMASK_32(9, 0)
+
+static SLIST_HEAD(, acc_device) hpre_list = SLIST_HEAD_INITIALIZER(hpre_list);
+
+struct hisi_qp *hpre_create_qp(uint8_t sq_type)
+{
+	struct acc_device *hpre_dev = NULL, *cur_dev = NULL;
+	uint32_t max_qp_num = 0;
+	uint32_t free_qp_num;
+	struct hisi_qm *qm;
+
+	SLIST_FOREACH(cur_dev, &hpre_list, link) {
+		qm = &cur_dev->qm;
+		free_qp_num = (qm->fun_type == HISI_QM_HW_PF ? HISI_QM_PF_Q_NUM
+			       : HISI_QM_VF_Q_NUM) - qm->qp_in_used;
+		if (free_qp_num > max_qp_num) {
+			max_qp_num = free_qp_num;
+			hpre_dev = cur_dev;
+		}
+	}
+
+	if (!hpre_dev) {
+		EMSG("No available hpre device");
+		return NULL;
+	}
+
+	return hisi_qm_create_qp(&hpre_dev->qm, sq_type);
+}
+
+int32_t crypto_bin_to_hpre_bin(uint8_t *dst, const uint8_t *src,
```

Comment:
> Please give a `hpre_` prefix to this function, perhaps `hpre_bin_from_crypto_bin()`. What is this function doing? Is it only copying and padding with zeroes or does it do something more?

Yes, it is only copying and padding with zeroes.
So that, it can translate the data format from the crypto_bin to the hpre_bin which required by the hardware.